### ISSUE TITLE
Evolving the SessionCache

### ISSUE THREAD

**Issue by boaks** [eclipse/californium#1345]
Though some issues around the `SessionCache`have been raised, it may be a good idea to start a discussion about this `SessionCache` and the intended usage.

Some related issues of the past as starting point:
#948 
#1163 
#1343 

From #1343:
The SessionCache seems for me to be not completely "specified". I don't know, if it is assumed to be "consistent" or "eventually consistent". The "removes" in case of a missing session id in the cache seems to point to the first, but intended "remote implementations" would then slow down too much. So, FMPOV, there is even specification work left, which then may change the implementations. That will be "time consuming".

I currently don't use this SessionCache and currently I prefer to improve the connection id usage in a cluster. Therefore my plan to invest time is more to make the "connection id cluster" working, than a clustered session resumption. With that it depends on others (including you :-) ), how much time you want to spend into the SessionCache.

**Comment by sbernard31** [comment]
> I don't know, if it is assumed to be "consistent" or "eventually consistent".

Could you explain what you mean by "consistent" or "eventually consistent" ?

> Therefore my plan to invest time is more to make the "connection id cluster" working

Could you describe what doesn't work for now ?
And what would mean exactly  "connection id cluster" working ?
Are you talking about a share/persistent connection store ?

**Comment by boaks** [comment]
> Could you explain what you mean by "consistent" or "eventually consistent" ?

["eventually consistent"](https://en.wikipedia.org/wiki/Eventual_consistency)

Generally I didn't invest too much time in understanding the `SessionCache`. One of the outcomes of this discussion may be, that it gets clear, which "consistent" model is used and how the connection store then uses the session cache (e.g. no remove in find).  

> Could you describe what doesn't work for now ?

There is simply no adoption of the [DTLS-1.2-connection-ID-based-load-balancer](https://github.com/eclipse/californium/wiki/DTLS-1.2-connection-ID-based-load-balancer). With that, I would like to experiment to forward the messages to the right node within scandium. e.g. check the cid, if it's not the own, forward the message to the right scandium node. Add the original source address to the udp-message, so that i java implementation is possible. That will not be that performant as a "external load balancer solution", but hopefully it works good enough to gain more attraction, more user, and so the chance to get the adoption for a real cid-load-balancer. 

> Are you talking about a share/persistent connection store ?

No, in my opinion the "record sequence numbers" will prevent such a implementation.

**Comment by sbernard31** [comment]
>  `forward the message to the right scandium node`

I don't know if we are representative, but if one day we use connection-Id, I'm pretty sure we will choose the load-balancer solution.

About a load-balancer dedicated to CoAP/LWM2M/Scandium use cases, [I'm still working on it based on XDP](https://github.com/AirVantage/sbulb).  It's a very young project but we begin to use it in production.

I'm currently working on ipv6 support but I still have the [connectionID support in mind](https://github.com/AirVantage/sbulb/issues/9). (but this is not consider as a priority by Sierra Wireless for now :-/)

[No license](https://github.com/AirVantage/sbulb/issues/13) was chosen for now but we want to make it open source.

If we want to make  `connection-ID`  more accessible, maybe we also need a [wireshark support](https://ask.wireshark.org/question/7618/support-of-dtls-connection-id/).

**Comment by sbernard31** [comment]
I read the ["eventually consistent"](https://en.wikipedia.org/wiki/Eventual_consistency) but still not crystal clear to me.

Anyway my 2 cents, 
We have a cluster Server based on Scandium/Californium behind some LVS or [SBULB](https://github.com/AirVantage/sbulb) load-balancer. We currently don't use `SessionCache` but we could use it to increase our number of abbreviated handshake (instead of full handshake).
The needs would be : 
 - share the session between nodes
 - persist session to keep it after failure or re-deployment.

My feeling is  this is acceptable to lost few session in cache meaning that some time we will not be able to do an abbreviated handshake.  About the remove, I understand it as : "This session MUST NOT be resumed anymore" (lifetime expire, credentials compromised, ...) and a remove must be safe/guaranty.

**Comment by boaks** [comment]
> About the remove, I understand it as : "This session MUST NOT be resumed anymore" (lifetime expire, credentials compromised, ...) and a remove must be safe/guaranty.

Or, the session is not already completely transferred to the cache. In that case, a connection will be accidentally removed.

**Comment by sbernard31** [comment]
> Or, the session is not already completely transferred to the cache. In that case, a connection will be accidentally removed.

I was talking about my needs as user, not how it behave currently.

**Comment by sbernard31** [comment]
As a user, 
Removing a session from the cache should only prevent to resume a session but not remove an existing connection. I talk about [DTLS concept](https://github.com/eclipse/californium/issues/1163), I know DTLS concept does not really match Scandium Classes)

**Comment by sbernard31** [comment]
Just to let you know, we begin to use `SessionCache` in a cluster environment in production.
First impression is pretty good. Thanks to it more sessions can be resumed (limiting the number of full handshake).

**Comment by sbernard31** [comment]
I should maybe ask this in another thread ? :think tell me if this is better.

Since we add `SessionCache`, we see a significant increase of CPU utilization. (from 10% cpu usage to 15%)
For now we have no real explanation of this.  We use `setVerifyPeersOnResumptionThreshold(0)`

Reading the code about SessionCache, I can not see what could generate this. I see a lot of Byte arrays.copy/clear (because of secretKey on SessionTicket.encode/decode) but I would be very surprised if this was related to that.

**Comment by erikmell** [comment]
> Just to let you know, we begin to use `SessionCache` in a cluster environment in production.
> First impression is pretty good. Thanks to it more sessions can be resumed (limiting the number of full handshake).

Would you like to elaborate on your setup using SessionCache. We have just recently set SessionCache up for redundancy/load balancing but still have some obstacles to fix.

1. To survive server restart we use the ClientSessionCache. This can take pretty long to start up when there is a lot of Sessions. What strategy do you use to clean sessions? For NAT:ed devices IP/port mapping will be outdated pretty quick but connectionId could make it last longer. For fixed ip:s the lifetime can be longer. But the sessionCache does not know when the session was used, so its hard to update/clear a TTL any thoughts on this?
 
2. Normally our loadbalancer will be sticky based on ip/port, if a node(1) goes down traffic will be routed to another node(2), for a new connection using the same session this works, but ongoing connections are broken. In get() there is a log "{}connection: missing connection for{}!". I guess that node(2) would need to sync with the cache before get to support this?

**Comment by boaks** [comment]
> but ongoing connections are broken. 

TLS/DTLS is based on two stage. The first negotiates the crypto parameter and the master secret. That's called session. The second contains then the keys and the record sequence numbers, that's called connection (TLS) or association (DTLS). For DTLS, the received record sequence numbers must be kept in a "window buffer".
The data in the session is negotiated on full handshakes and are long lasting. The data in the association is split into middle lasting (keys) and short lasting (record sequence numbers), they change with every exchange records.

In Californium the data is split into the `DTLSSession` and the `DTLSConnectionState`. "Historical" (and maybe changing in a future version :-) ), the `DTLSSession` maintains also the record sequence numbers.

Californium supports a `SessionTicket` ( a serialized session), but that doesn't contain the association state nor the record sequence numbers. Also some of the later added crypto parameter are missing, I hope, I can add them with a 3.0.

Currently you can just persist the session, which requires a "resumption handshake" to get the association data. To persist the association data, I consider in the future a graceful shutdown in order to be able to save the "record sequence numbers" (they must not change afterwards). But for now, it's not implemented. If you consider/plan to implement it, I would welcome a contribution. But even an experience report will be really helpful.

> This can take pretty long to start up 

Can you give us a impression, how many session take which time? My feeling is, this may be more related to the persistent mechanism.

**Comment by erikmell** [comment]
> Currently you can just persist the session, which requires a "resumption handshake" to get the association data. To persist the association data, I consider in the future a graceful shutdown in order to be able to save the "record sequence numbers" (they must not change afterwards). But for now, it's not implemented. If you consider/plan to implement it, I would welcome a contribution. But even an experience report will be really helpful.

Do you think It could be feasible to experiment with loading the session from SessionCache if the get() on a failover-node does not find a local matching connection, forcing a resumption handshake? 

> > This can take pretty long to start up
> 
> Can you give us a impression, how many session take which time? My feeling is, this may be more related to the persistent mechanism.
When I experienced problems we had about 120000 sessions. On the server it took 12+hours, but I have not tracked the problem down yet, when loading from a local copy of the DB (redis) it only took a couple of minutes so I guess the problem is within the server environment and not an californium issue.

**Comment by boaks** [comment]
> Do you think It could be feasible to experiment with loading the session from SessionCache if the get() on a failover-node does not find a local matching connection, forcing a resumption handshake?

Your dtls-client has three possibilities:
- starting a full handshake (without available valid session ID)
- starting a resumption handshake (with valid session ID). The dtls-server may refuse the resumption and switch into a full handshake.
- sending an application record. The dtls-server either decrypt the record successfully, or drops it silently.

I'm not sure, if you want to use a DTLS role exchange (means the server will act as dtls-client). I guess not. If not, it's not the dtls-server, which chose to resume a session, it's the dtls-client. The dtls-server can only refuse that. But even more, without a handshake from the dtls-client, the dtls-server is intended to drop that message!

The "session cache" is only involved, if a dtls-client want's to resume such a session. Otherwise the resulting session is just stored their. You chose to use the `ClientSessionCache` for the dtls-server. But that will only be able to restore connection with available `SessionTickets`. The intended usage is a dtls-client, which use that ticket to resume a session. On the dtls-server side, this doesn't work. For that dtls-server side the `SessionCache` is used to lookup for sessions, only when a dtls-client wants to resume such a session.

I assume, you want a scenario, where a "application record" is received by a failover-node. The would require a really "high frequently update of the association states (record sequence numbers). FMPOV, that's not really performant. The "load" to execute a coaps request is such small, that the load for sync the states would be relevant. It may be also a wrong assumption, that other technique works like that (that is rare). Basically the most detect the failure and (re-)connect to the failover-node.

> When I experienced problems we had about 120000 sessions. On the server it took 12+hours, but I have not tracked the problem down yet, when loading from a local copy of the DB (redis) it only took a couple of minutes so I guess the problem is within the server environment and not an californium issue.
 
I would wonder, if loading 120000 session tickets would take more than a few seconds for californium.
But @sbernard31 has sure more experience on what to expect from redis.

**Comment by boaks** [comment]
By the was: the javadoc of the `SessionCache` and  `ClientSessionCache` are misleading and not really maintained for long. Too much other work ...

**Comment by sbernard31** [comment]
Some explanation about how and why we use it.
We are deploying several server (Leshan-LWM2M) in cluster behind loadbalancer. 

The use case is mainly about  devices behind NAT (dynamic ip environment).
Unfortunately, we are not using **connection id**, and so to resolve this issue all devices start communication by a handshake.
A full one if device has no session else an abbreviated try.

Without `SessionCache`, session are not persisted and not shared meaning : 
- in case of server reboot, all "abbreviated handshake try" will turn into "full handshake"
- in normal use case, you have 1/x chance (x is the number of instance in the cluster) to succeed your abbreviated handshake.

The idea to use `SessionCache` is just to maximize the number of abbreviated handshake (or more precisely to decrease the number of abbreviated handshake which will turn into full handshake)

I guess the benefits should be : 
- less resources usage at device side. (cpu ? bandwith?)
- less resources usage at server side (cpu ? bandwidth?)
(for bandwidth this is true mainly for RPK/x509 not so true for PSK)

On our side we did that because of some devices which behaves badly when abbreviated turn into full handshake...
We expected also some cpu benefits at server side => currently this is pretty much the opposite :sweat_smile: but anyway devices are happier now. 

We are using redis and implements only `SessionCache`(not `ClientSessionCache` as currently we don't care about initiated abbreviated handshake at server side).
Our keys expire after a while (something like 7days but could be less in the future) without r/w access. If session is lost, it's not a big deal as we see that as an optimization. I mean the device must support the case where the session doesn't exist anymore.
To avoid to create too many Session in store, when a session is created we remove previous session from the same peer. (in our case we use PSK ID as peer identifier)

Our `RedisSessionCache` could be added in Leshan at midterm.

**Comment by boaks** [comment]
> => currently this is pretty much the opposite sweat_smile 

It's mainly a question, how much cpu a full handshake takes more than a resumption handshake, compared to the "remote call" of the session cache.
As you wrote, it's more for the clients, than for the performance.

**Comment by sbernard31** [comment]
I tried to do some tests about this.
**My current procedure :**
Using `sc-dtls-example-client`, I modified it to do a handshake (full or abbreviated) for each message sent.
Using `sc-dtls-example-server`, I modified it to use RedisSessionCache (or not).
I use only PSK. DTLS client, server and redis run on the same machine.

I did 4 runs for each case and ignore result of first run.
Each run, I send 1 client which send 30000 messages. (each message = 1 handshake)

As result, I used message per seconds returns by `sc-dtls-example-client` and cpu cummulated usage (TIME) from `ps` command.

(I didn't find a best way for now :thinking:  )

**The result:**


Full Handshake | message per s | cpu TIME | cpu TIME/run
-- | -- | -- | --
1rst run | 1522 | 00:24:00 |  
2nd run | 1835 | 00:36:00 | 00:12:00
3rd run | 1765 | 00:50:00 | 00:14:00
4th run | 1898 | 01:02:00 | 00:12:00
mean (without 1st run) | 1833 |   | 00:12:40

Abbreviated Handshake | message per s | cpu TIME | cpu TIME/run
-- | -- | -- | --
1rst run | 2262 | 00:19:00 |  
2nd run | 2514 | 00:28:00 | 00:09:00
3rd run | 2503 | 00:38:00 | 00:10:00
4th run | 2913 | 00:45:00 | 00:07:00
mean (without 1st run) | 2643 |   | 00:08:40

Abbreviated + session Cache | message per s | cpu TIME | cpu TIME/run
-- | -- | -- | --
1rst run | 1381 | 00:24:00 |  
2nd run | 1650 | 00:36:00 | 00:12:00
3rd run | 1555 | 00:54:00 | 00:18:00
4th run | 1594 | 01:07:00 | 00:13:00
mean (without 1st run) | 1600 |   | 00:14:20

I tried a little optimization where I check is session exist before to put it in Redis.
Abbreviated + session Cache () | message per s | cpu TIME | cpu TIME/run
-- | -- | -- | --
1rst run | 1537 | 00:23:00 |  
2nd run | 1688 | 00:36:00 | 00:13:00
3rd run | 1650 | 00:51:00 | 00:15:00
4th run | 1885 | 01:02:00 | 00:11:00
mean (without 1st run) | 1741 |   | 00:13:00

"Message per s" and "cpu TIME/ run" sounds inversely proportional which makes sense :thinking: 

I had  a look with visualVM profiler, and hotspot seems to be around put/get access of sessionCache.
In particular, all redis call.
About Deserialization/Serialization this sounds less impacting than redis call but decoding SessionTicket sounds more expensive than encoding it.
This sounds to confirm previous result.

There is maybe some solution to improve this :
- Put session in cache only for full handshake. (If I well understand session does not change after resume so we don't need to add it again ?)
- Do not try to look in SessionCache if session is already in InMemoryConnectionStore (this changes the behavior a bit meaning that removing from SessionCache does not guaranty that the session is totally removed  => This is opposite to my previous "needs description" but maybe it make sense if it improves performance)
- we could add a `contain` method to SessionCache which could avoid some useless TicketSession decoding.

**Comment by boaks** [comment]
> Put session in cache only for full handshake.

I'm not sure, if such sessions then may expire. But that may depend more on the intended session-management of the cache.

> we could add a contain method to SessionCache which could avoid some useless TicketSession decoding.

Sure, but above you wrote, that it is "less impacting", so I wouldn't expect a large improvement.

I guess, if you adapt your test with a certain amount of application records and then do the handshakes in parallel, then you may also see the "blocks", as mentioned in #1343 

So FMPOV, one of the important decisions is, if the PUT/GET/DEL of the connection store must be in atomic with these operation in the session-cache, or if they could be decoupled. But, as I wrote at the begin, I still busy with other stuff. So, everybody is welcome to help.

**Comment by sbernard31** [comment]
Here is my current understanding.

Firstable, wording is bit confusing because in fact `SessionCache` is a kind of **Persistent/Shared Session Store** and `InMemoryConnectionStore` is a **connection store + a kind of local session cache**.
So below I will use store(SessionCache) and local cache(map in ConnectionStore) wording.

If we want to focus on performance : 
1) We should look in store only if value is not in local cache. (this will decrease drastically the access to store)
2) Local cache should expire some time to force to recheck value in store.

In fact all of this could be achieve in a `SessionCache` implementation with an internal map as cache + requesting redis store when needed. But the internal map will be a kind of duplicate of local session cache in InMemoryConnectionStore.

So I'm currently thinking that maybe a right way to do will be to : 
- Remove the local session cache in InMemoryConnectionStore.
- provide a default inMemorySessionCache.

> PUT/GET/DEL of the connection store must be in atomic with these operation in the session-cache, or if they could be decoupled.

"Decoupled" means executed in another thread ? 
For GET, I'm not sure it's worth creating an async API. (especially if we have a local cache which prevent access to store)
For PUT/DEL, If think it could be decoupled but maybe it depends of use cases.
Note that this decoupling (for PUT/DEL) could be done internally in `SessionCache`, I mean implementation can create new task to put/delete data in store by its own thread pool. (Or I missed something?)

**Comment by boaks** [comment]
> For GET, I'm not sure it's worth creating an async API. 

No, but a async PUT will change the behavior of the GET. At least the current implementation, without "caching" the GET, will behave different, if the PUT is async, because a new GET may have a different result, if the PU is not completed.

> could be done internally

Yes, but it changes the details of the GET and maybe DEL (see above).
So define a async API would in my opinion also need to define the interaction of both.

For changing the names:
I think, that after a 2.5 starting over with a 3.0 would make sense, because I would like to get rid of the deprecates. That would be then also a chance to adapt the names.

**Comment by boaks** [comment]
Just for information:

The #1424 gets more and more available.

[built-in CID-load-balancer](https://github.com/eclipse/californium/tree/wip_dtls_cid_cluster/demo-apps/cf-extplugtest-server#built-in-dtls-connection-id-load-balancer-cluster)

Feedback is very welcome.

**Comment by erikmell** [comment]
> 
> Currently you can just persist the session, which requires a "resumption handshake" to get the association data. To persist the association data, I consider in the future a graceful shutdown in order to be able to save the "record sequence numbers" (they must not change afterwards). But for now, it's not implemented. If you consider/plan to implement it, I would welcome a contribution. But even an experience report will be really helpful.

Thank you all for your detailed answers and all the effort you put in to Californium. We are about to release a series of battery powered devices with with lifetime >7-10 years. Hence the need for service-windows on the server side with graceful shutdown restart come up, Every handshake is shorter lifetime for these devices. It is very likely that we will start laborate on this quite soon. Should I file a separate ticket to continue discussions on that? I'm pretty certain we will get a OK from our managers to make a contribution.

**Comment by boaks** [comment]
> to make a contribution.

Welcome.

Just to mention:
If your device comes up with "different" ip-endpoints (address:port) at the server, you may need [DTLS CID](https://www.ietf.org/archive/id/draft-ietf-tls-dtls-connection-id-07.txt) implemented on the client as well.
That is required even without graceful-startover just in order to prevent from handshake after a quiet period.
See [Bypassing-NATs](https://github.com/eclipse/californium/wiki/DTLS-1.2-connection-ID-bypassing-NATs).
  
My current plan was to start over with 3.0 after the 2.5.0, removing the deprecates and then clean up the Connection/DTLSSession/DTLSConnectionState in order to prepare for the graceful-shutdown-restart.
My feeling is, that will take until Christmas (the preparation not the graceful-startover :-) ).
I'm not sure, if a faster timeline is realistic.

**Comment by erikmell** [comment]
> If your device comes up with "different" ip-endpoints (address:port) at the server, you may need [DTLS CID](https://www.ietf.org/archive/id/draft-ietf-tls-dtls-connection-id-07.txt) implemented on the client as well.
> That is required even without graceful-startover just in order to prevent from handshake after a quiet period.
> See [Bypassing-NATs](https://github.com/eclipse/californium/wiki/DTLS-1.2-connection-ID-bypassing-NATs)

Thank you, we have CID enabled on the devices using mbed tls. We have both scenarios, devices on private-apn with static ip, and devices coming via public APN:s that frequently changes IP.

**Comment by boaks** [comment]
For the "graceful shutdown" I prepared a first PoC in PR #1468 

Checkout that PR, the "demo-apps/cf-extplugtest-server" contains a section `Benchmarks - Graceful Shutdown` with instructions. That very simple demo just serializes all connections and removes them from the connector. afterwards they can be loaded again into the connector.

**Comment by boaks** [comment]
I started to test some ideas about the "distributed session cache":

full handshake:
- after success, transfer the session (async PUT). 

resumption handshake:

- same node/session may be found locally. FMPOV, a (async GET) request is required to ensure, that the session is still valid (means not aborted by an alert). That requires also the "state" of the transfer of that session on the full-handshake.

- session is not found locally, then it requires a (async GET) request.
    
With that, I think, deployments, which use a "distributed" session store,
MUST always use a HelloVerifyRequest, because they always require to do
an potential "expensive" GET request.

(async) requests are required, to relax the thread pool (reduce the influence of handshakes according other appl-records). 



  

Alternative it may be assumed, that there are no alert ... (I don't feel to comfortable with that).

**Comment by boaks** [comment]
Some questions about corner-cases:

Full-handshake:
- successful, but the transfer of the session fails. What should happen?

**Comment by sbernard31** [comment]
> successful, but the transfer of the session fails. What should happen?

It seems to me that persisting the session is just best effort if we lost some It's OK. I mean when this happen this is just a full handshake instead of abbreviated one. 

Currently, I have more concern about how to be able to remove all connection + session for given compromised credentials in a way we can be sure there is no race condition. (I didn't think too much of this, but I feel that we can maybe face some problem here ? :thinking:)

> same node/session may be found locally. FMPOV, a (async GET) request is required to ensure, that the session is still valid.

Please read this https://github.com/eclipse/californium/issues/1345#issuecomment-700732372 again (if this is not already done :grin:), I feel we should maybe change this idea about local cache / Session cache and replace this by a `SessionStore` with a default in memory implementation and so it's up to the user to create a persistent one with eventually local memory cache.  This way it's up to him to check or not in persistent store when there is something in local memory cache. (I hope this is not so confusing)

**Comment by boaks** [comment]
> `Session cache` and replace this by a `SessionStore` with a default in memory implementation.

FMPOV, the current `InMemoryConnectionStore` should be at last very close to that.
It is possible to have connections only with a session or ticket, but no address. There maybe some unseen pitfalls, but generally, that should work.

**Comment by boaks** [comment]
> Currently, I have more concern about how to be able to remove all connection + session for given compromised credentials in a way we can be sure there is no race condition. (I didn't think too much of this, but I feel that we can maybe face some problem here ? thinking)

FMPOV, it usually takes a long time until such compromised credentials are detected. With that, it's more important, that "over the time", all connections in the "cluster" are removed. But maybe I misinterpret the "race condition".

**Comment by sbernard31** [comment]
> FMPOV, the current InMemoryConnectionStore should be at last very close to that.
It is possible to have connections only with a session or ticket, but no address. There maybe some unseen pitfalls, but generally, that should work.

I'm not sure if you agree or not with my proposal ?

**Comment by sbernard31** [comment]
> FMPOV, it usually takes a long time until such compromised credentials are detected. With that, it's more important, that "over the time", all connections in the "cluster" are removed. But maybe I misinterpret the "race condition".

I try to explain a bit more. :-)
Imagine you don't want to allow a given credential to be used anymore (E.g. a given psk ID because it is compromised or just changed)
Currently my idea (maybe I'm wrong), is to remove all in memory connection in the cluster and remove session in share store.
My concern is that maybe as those 2 operations are not done at same time, we could face some race condition which could end with some session or connection which are not deleted.

Eg. I remove connection, then I remove session. (but between a new connection was created)

**Comment by boaks** [comment]
> I'm not sure if you agree or not with my idea ?

I don't see anything, which holds someone up from trying that.
Therefore my hint, that the current implementation should already be able to store such "sessions only". 
Or do I oversee something?

**Comment by sbernard31** [comment]
I think about maybe removing  : 
`protected final ConcurrentMap<SessionId, Connection> connectionsByEstablishedSession;` from `InMemoryConnectionStore`
This way when we resume a session we only search in `SessionStore` (or `SessionTicketStore`) and no more in this local store :point_up:.

The idea, user implements its `SessionStore` and use local memory cache if needed and chose this local cache behavior.

(This refactoring could maybe be a chance to also refactor Connection/DTLSession class in a way it better reflect DTLS concept : https://github.com/eclipse/californium/issues/1163#issuecomment-564692966 by maybe removing DTLSSession)

So a `connectionStore` which store "dtls connection" and a `SessionTicketStore` which stores `SessionTicket`.

**Comment by boaks** [comment]
My feeling is, remove the "SessionCache" and try out, if the current connection store does it.

**Comment by sbernard31** [comment]
I'm lost I have no idea how this could be possible with current connection store ? Unless I don't get what you call "current connection store".

**Comment by boaks** [comment]
My understanding of such a in memory solution:
If a session gets established, transfer that to other nodes.
On other nodes, add the transferred session to the connection store.
My point is, if it's a "direct in memory solution", the sychronization seems to be critical and so I would remove the SessionCache callbacks.

**Comment by sbernard31** [comment]
I think you don't understand me at all and I don't know how to explain it better :-/
My idea is about refactoring the code to make all those concept clearer and allow following usecase :
1. in memory sessionTicketStore (default)
2. persistent SessionTicketStore (for failover)
3. shared persistent SessionTicketStore (for cluster)
4. shared persistent with local memory cache (for cluster and optimized access to shared store)

(unless it's me who doesn't understand you :sweat_smile: )

**Comment by boaks** [comment]
My idea would be,
Start with a PoC for such a in memory session store. Demonstrates, that it works to distribute the session ahead.
To do that "fast", I would just use the connection store. It possible to store connections with tickets or sessions only there. At least, I think it should be possible.
I don't think, that define (theoretical) APIs first works better. At least not in my experience.

**Comment by boaks** [comment]
Let me add:

The idea of a "pre-distributed session in memory solution" seems to be the simplest, if it works.

If not, my impression so far is, that a "async remote API", will not really fit into that session cache nor connection store. Such a API requires a very different redesign. I started to think on also mark session ids with node-ids (similar to the cid), and then just on resumption try to load the session from such a specific node (using the introduced cluster internal communication). Such an approach will unfortunately require a lot of development and test time.

**Comment by sbernard31** [comment]
>  I don't think, that define (theoretical) APIs first works better. At least not in my experience.

Clarify concept, separate concern and define API first is how I work on Leshan trying to keep the code as understandable as possible.

**Comment by boaks** [comment]
@erikmell 

PR #1486 gets more and more "usable". 
If you're interested in first experience, your feedback will be welcome.

**Comment by erikmell** [comment]
> @erikmell
> 
> PR #1486 gets more and more "usable".
> If you're interested in first experience, your feedback will be welcome.

Sorry for not answering in a long time. Had a long christmas leave and after coming back there's been too much to do and some organisational changes. 
This will definitely be a good thing for our battery powered NbIoT devices, not forcing a handshake after server restart.

**Comment by boaks** [comment]
The "wip feature" is on `master`.

Some docu how to use it for the extended plugtest demo server is here:

[Benchmarks - Graceful Shutdown](https://github.com/eclipse/californium/tree/master/demo-apps/cf-extplugtest-server#benchmarks---graceful-shutdown)

If you want to test it on your own server's, [CoapServer.saveAllConnectors](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java#L488) is a good point to start.

Just to mention:
It's wip, this means especially, californium may change the data format. As long as you use it for updates of your parts keeping Californium at the "commit", you should be able to use it.

**Comment by boaks** [comment]
The blue/green update with DTLS graceful restart is available for testing!

Checkout PR #1547 !

**Comment by boaks** [comment]
@sbernard31 

The master (upcoming 3.0.0-M1, when ever :-) ), contains now the redesigned DTLS-state. The previous `DTLSSession` is split into `DTLSSession` and `DTLSContext`. I'm still not sure, if renaming "connection" into "relation" really clarifies more than it mixes up, so I kept that term "connection".

With that, it seem for me:
- if a "session cache" (or renamed to "SessionStore") is used, the dtls-connection-store may use that instead of the own `connectionsByEstablishedSession` (the change in the implementation is pending).
- if a "session cache" is **NOT** used, I would prefer to keep using the current approach with `connectionsByEstablishedSession`.

Therefore I think, removing the `connectionsByEstablishedSession` is too much, but disable the use, is a "Session Cache/Store" is used, that should work.

FMPOV, with the current interface `SessionCache` (maybe renamed to Stor), it should be possible to implement the solution you proposed. On "established" would then require not only to store it, it mus also deploy it to other nodes. 

So:
- Rename SessionCache into SessionStore?
- disable the use of `connectionsByEstablishedSession`, if the SessionCache/Store is provided?

**Comment by sbernard31** [comment]
> The previous DTLSSession is split into DTLSSession and DTLSContext

:+1: 

>  I'm still not sure, if renaming "connection" into "relation" really clarifies more than it mixes up, so I kept that term "connection".

I understand this could be a lot of renaming.
What currently make the current `Connection` concept from Californium differ from the DTLS concept ?
Regarding https://github.com/eclipse/californium/issues/1163#issuecomment-564692966, I guess that : 
- DTLS Session concept matches  `DTLSSession`
- DTLS connection concept matches  `DTLSContext` ? 
- and `Connection` is a container for information of the relation which is not so clear to me. 

> Therefore I think, removing the connectionsByEstablishedSession is too much, but disable the use, if a "Session Cache/Store" is used, that should work.

Considering :   
_"if a "session cache" (or renamed to "SessionStore") is used, the dtls-connection-store may use that instead of the own connectionsByEstablishedSession`connectionsByEstablishedSession`"_
If we provide an `InMemorySessionStore` we should have the same behavior than without `SessionStore` at all. 
If this is true then we should be able to remove `connectionsByEstablishedSession` attribute from `InMemoryConnectionStore` ? 
I guess I totally missed something ? :sweat_smile: 

>  On "established" would then require not only to store it, it mus also deploy it to other nodes.

I don't get this part. 

> Rename SessionCache into SessionStore?

If you agree this is more a store than a cache, I think renaming makes sense.

> disable the use of connectionsByEstablishedSession, if the SessionCache/Store is provided?

I sill think the `connectionsByEstablishedSession` should be deleted but probably because I missed something.

**Comment by boaks** [comment]
Basically, I prefer to have the "SessionCache/Store" optional. With that, I don't plan an implementation of that interface. And therefore keeping "connectionsByEstablishedSession" as "default" is for me the preferred way.
(An alternative would be to add a simple in-memory-hashmap based implementation by default, if no other is provided.)

>    On "established" would then require not only to store it, it mus also deploy it to other nodes.
> I don't get this part. 

Hm, I considered that as the solution you proposed. If not, simply ignore it.

**Comment by boaks** [comment]
> If you agree this is more a store than a cache

The difference of these terms is for me:
- a store has similar access times, if the same items are requested.
- a cache may speed up, if the same items are requested.

Though a "time consuming get" would require a change in the synchronization scope (with that, the issues started), I think, a "cache" behavior is not , what should be the defined behavior.

**Comment by sbernard31** [comment]
> An alternative would be to add a simple in-memory-hashmap based implementation by default, if no other is provided. 

I feel it would be better. This way there is only 1 behavior and just several implementation, it seems more straightforward to me.
I can not the the benefits of the other alternative(keeping `connectionsByEstablishedSession` attributes) ? Do you have a specific one in mind ?

**Comment by boaks** [comment]
> Do you have a specific one in mind ?

Yep, it doesn't require to use the `SessionTicket`. And the current `InMemoryConnectionStore` takes also care about cleanup old connections based on the session id. (I'm not sure, how that was intended by the idea of that Session??? interface.)

**Comment by sbernard31** [comment]
I looked at this a little bit.

Regarding `SessionTicket`, I see that SessionCache API is not so consistent. 
```java
// current
void put(DTLSSession session);
SessionTicket get(SessionId id);
```
I would rather go to : 
```java
// 1. alternative
void put(SessionId id, SessionTicket ticket);
SessionTicket get(SessionId id);
```
OR if we want to avoid `SessionTicket`<-->`DTLSSession` conversion for in memory implementation. (like the default one or persistent one with in memory cache)
```java
// 2. alternative
void put(SessionId id, DTLSSession session);
DTLSSession get(SessionId id);
```

So either your first remark about `SessionTicket` is not so important and we could go for 1.
or we can take it into account with the cost of more documentation in `SessionStore` explaining this is only needed to persist `SessionTicket`.  (or we add an `AbstractPersistentSessionStore` which takes in parameter a `SessionTicketStore`)

**Comment by sbernard31** [comment]
>  the current InMemoryConnectionStore takes also care about cleanup old connections based on the session id. 

This is the lifetime/lifecycle question about `Session`. My intuition is that the API should allow to implement most used way to handle this. (captain obvious ... :sweat_smile:)

Some classic way I see is :
 - having a LRU cache with a max size. 
 - LRU cache with max size + expiration lifetime.
 - external remove because of compromised credentials
 
I don't know exactly what is the current behavior, but either it should be replaced by one those above :point_up:  or the new API should be allow to implement it ? :thinking: 

Which kind of clean-up exactly you think it will not be able to do with the `SessionStore` API ?

**Comment by boaks** [comment]
[putEstablishedSession](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/InMemoryConnectionStore.java#L359)

At least, I don't see a obvious solution without the `connectionsByEstablishedSession`.

**Comment by sbernard31** [comment]
This is about cleaning previous connection for the same session. To be sure there is only 1 connection by session, right ? 

I guess at (D)TLS side, there is nothing to prevent to have several connections for the same session ? (see [conccurent connection](https://security.stackexchange.com/questions/77521/do-concurrent-connections-to-an-https-website-perform-separate-ssl-handshakes)). So unless there is this kind of restriction at CoAP side and we want to limit Scandium DTLS implement to the CoAP flavor, I would just don't remove the connection. 

It should not be an issue as Connection will be removed later anyway thanks to LRU cache or I missed something ?

**Comment by boaks** [comment]
> I guess at (D)TLS side, there is nothing to prevent to have several connections for the same session ?

No, it's not forbidden. It makes just not that much sense. Similiar to renegotiation. Both seems in my opinion to be more caused by TCP than by the encryption.
My point is more, that everyone must care about that and check, if it doesn't harm. It changes the behavior of the last 2-3 years. Therefore I would prefer, that those who want to use the "distributed session store", are those who then report their experience with it.

**Comment by sbernard31** [comment]
As far as I understand this will not change the "behavior", just the memory management. (not so true this change the behavior a little as  it will allow several connection for one session but it should not harm)
But If you are not so confident and prefer to let user go back to the previous behavior, I would keep the `connectionsByEstablishedSession` attribute and provide the `InMemorySessionStore` as default behavior.

So if someone complain we can propose to go back to previous memory management and if after a while no complain we could maybe consider to remove it ?

**Comment by boaks** [comment]
> keep the connectionsByEstablishedSession attribute and provide the InMemorySessionStore as default behavior.

Hm, that would require some additional changes, because:

```
public DTLSConnector(final DtlsConnectorConfig configuration, final SessionCache sessionCache)
```

Using `null` will create the default, using non-`null` will use a special implementation. So, I'm not sure, how it will be possible to go back to the `connectionsByEstablishedSession`.

**Comment by sbernard31** [comment]
I guess with a new setter which activates the old behavior ? `setDoNotUseSessionStore(true)` ?

**Comment by sbernard31** [comment]
Uups I just get that you are talking about `DTLSConnector`.
My previous idea was that `SessionStore` was set via `DtlsConnectorConfig` and currently this is not possible right ?
But maybe we should ? :thinking:

**Comment by boaks** [comment]
> But maybe we should (being able to set SessionStore via DtlsConnectorConfig.) ?

That may be a solution.

**Comment by boaks** [comment]
One resulting change in the "SessionStore" LRU behavior will be the "U". Currently it's the time of the last exchanged message, with the "SessionStore" it will be the time of the last resumption.

**Comment by sbernard31** [comment]
Good point !

I don't know if this is an issue.
To be honest, I don't even know what should be the right default behavior.
I mean when the store is full which one should be removed first ? :thinking: 
 Maybe the session which was create in first ? so you avoid to keep it for a too long period in the store. I'm not sure. (This is again the question : "is there security issue to keep Session for too long")

Anyway IF we want to be able to keep the session regarding the last exchanged message THEN we could maybe add a new method to the store, like : 
- `void messageExchanged(SessionID [or] DTLSSession)`
- or if we want more detailed behavior : `void sessionUsed((SessionID [or] DTLSSession, Usage)` where `Usage` is an enum with different kind of event. (message sent, message received,  close_notitfy, ... )

**Comment by boaks** [comment]
I started yesterday evening with some code changes, hope I can finish that today. 

May first results:
without the "connectionsByEstablishedSession" some unit test are failing.
Some because the tests checks the "remaining in the connection store". Though that change is expected, the test must be adapted.
One other failure makes me more wondering. It's

```
DTLSConnectorAdvancedTest.testResumeWithVerify
```

With that, it requires the reanalyze, when a hello verify request should be used and when not.
At least, before removing that `connectionsByEstablishedSession`.

**Comment by boaks** [comment]
The 

```
DTLSConnectorAdvancedTest.testResumeWithVerify
```

was failing, because the `pendingHandshakesWithoutVerifiedPeer` was only used for "local connections". NO it's fixed and the other tests, which have checked the connections in the connection store, now check the sessions in the session store.

**Comment by boaks** [comment]
See PR #1550 for first results.

**Comment by boaks** [comment]
About the `connectionsByEstablishedSession`:

The difference in keep it is:
- remove connection on failing find.
- remove connection on establish session.

FMPOV, though the rest of the functionality not longer depends now on that  `connectionsByEstablishedSession`, I prefer to **not** use a `SessionStore`by default.
That limits this effect only to those, who want to use such a `SessionStore`.

**Comment by sbernard31** [comment]
_"remove connection on establish session"_ is about https://github.com/eclipse/californium/issues/1345#issuecomment-783541169 ?

_"remove connection on failing find"_,  I'm not sure to get this one. :thinking: 
Is it this part of the code ? https://github.com/eclipse/californium/blob/e7bd4e4bfe3076eb4f0271e788f68f836264b14e/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/InMemoryConnectionStore.java#L404-L415

**Comment by boaks** [comment]
Yes. Without the `connectionsByEstablishedSession`, it's not efficient to get the connection for a session. 

Both points are not "too" hard nor "show stopper". It's more my experience of the past months/years, that "forcing" the usage of the "SessionStore" will cause a couple of weeks testing on my side, which I'm not really want for now. The related PR cleans up a lot for resumption handshake (I'm not completely ready). So the left difference with using a "SessionStore" by default or not, seems for me to ensure, that the "SessionStore" is really working, which requires that time. So, FMPOV, those who want the "cluster resumption" should have a very solid base to verify, that it works or what needs to be changed.

**Comment by sbernard31** [comment]
To those 2 points : 
-  remove connection on failing find.
-  remove connection on establish session.

We should also add "the meaning of 'used' for LRU cache of SessionStore" (see https://github.com/eclipse/californium/issues/1345#issuecomment-784950197)

About keeping `connectionsByEstablishedSession` as default, I think it could make sense at very short term but at short or mid term we should go to the InMemorySessionStore as default to get feedback. 

I looked at the PR quickly (but not too deeply because there is several changes not really related to the `SessionStore` behavior and so not so easy to me to split it in my mind :sweat_smile:)
Anyway, I didn't see change on `SessionStore` API (like https://github.com/eclipse/californium/issues/1345#issuecomment-783501810 or https://github.com/eclipse/californium/issues/1345#issuecomment-784950197), this is something you consider to add for the first version or you prefer to not change it in a first time or you didn't thing about that or maybe you consider this is no so good idea ?

**Comment by boaks** [comment]
> Anyway, I didn't see changes on SessionStore API

Basically I spent my time in understand all the obscure "if/else/maybe" parts in order to see, what is really mandatory and not. That results in a medium cleanup of the "new client hello". One "major" consequence is the `ResumptionSupportingConnectionStore.find(SessionId)`, which now returns a `DTLSSession` instead of the `Connector`.

I'm still on verifying the changes so far. I will also try to test, if switching from `SessionTicket` to `DTLSSession` in the `SessionStore`. really makes sense. Maybe it just doesn't payoff ;-).

**Comment by boaks** [comment]
(I still feel, that a real, well-tested solution of this topic requires much more time, I'm able and willing to spend for now.)

Trying to follow all proposals and questions here, make me feel, that even that requires too much time. Anyway let me try to give some answers.

> To be honest, I don't even know what should be the right default behavior.
> I mean when the store is full which one should be removed first ? thinking
> Maybe the session which was create in first ? so you avoid to keep it for a too long period in the store. I'm not > sure. (This is again the question : "is there security issue to keep Session for too long")

I have also no preference here. For the current simple demo implementation `InMemorySessionStore`I choose the last put. For me, this is more similiiar to the used behavior then the age. Anyway, I guess it will require real experience to get a better approach.

> Anyway IF we want to be able to keep the session regarding the last exchanged message THEN we could maybe add a new method to the store, like :

I don't think, that "fine grained updates" should be used. But also here, I guess it will require real experience to get a better approach.

**Comment by boaks** [comment]
Working on PR #1555 makes me finally believe, that just removing the `SessionTicket` because it gets obsolete, is the best choice. The `SessionTicket` is mainly a copy of the `DTLSSession` (with final fields). And so, without using it really as `SessionTicket` to keep the data on the client side, that class doesn't make sense to be maintained. If such a SessionTicket gets used for that, it will be much easier to split the `DTLSSession.write/read` functions than to maintain it as separate class.

**Comment by sbernard31** [comment]
Does it impact the encoding decoding way ? 
- org.eclipse.californium.scandium.dtls.SessionTicket.encode(DatagramWriter)
- org.eclipse.californium.scandium.dtls.SessionTicket.decode(DatagramReader)

**Comment by boaks** [comment]
Sure, the point is, that any new field will change the encoding/decoding.  The `SessionTicket` 2.6 was therefore hard to maintain. The new idea about serialization is to start with a "version", which helps to overcome a situation, if a change is hard to support it in a backwards compatible way. With that it is possible to keep "old readers" but use "new writers".

**Comment by boaks** [comment]
If your question targets how to migrate "existing binary sessionticket v2.6", I can try to add that encoding to the `DTLSSession.readFrom()`.  That would also require to adapt the setter for the session id, but that should be not to hard.

**Comment by sbernard31** [comment]
> if your question targets how to migrate "existing binary sessionticket v2.6"

This was the idea behind question.
But more to be aware of the impact than to discourage you to not make that change. (I mean I'm generally not against refactoring to make the code better)

> I can try to add that encoding to the DTLSSession.readFrom(). That would also require to adapt the setter for the session id, but that should be not too hard.

If it's possible, I guess having backward compatibility solution could be good.

**Comment by boaks** [comment]
I tried to read the old encoded session tickets. But that requires also the old `PrincipalSerializer` and `ServerNames`, so I'm not sure if to add that pays off.

**Comment by boaks** [comment]
@sbernard31 

I added a second commit with the "backwards compatible session ticket" read.
Please have a rough check on that second commit, and signal, if you think, it paysoff. Otherwise I would remove that second commit.

**Comment by sbernard31** [comment]
I quickly look at it. 
To be honest I didn't think too much about how we will handle the migration but I guess we will need something like this. 
It's hard to me to guess if we should keep this it californium or not.
If you decide to not keep, could you create a branch containing this commit ? I guess that could be useful later at least as a code snippet.

**Comment by boaks** [comment]
> If you decide to not keep, could you create a branch containing this commit ? I guess that could be useful later at least as a code snippet.

FMPOV, that code is mainly executed "once". I'm not sure, what the best approach will be. At least a separate branch should help those, who want to migrate serialized 2.6 `SessionTicket`s.

[read_sessionticket](https://github.com/eclipse/californium/tree/read_sessionticket)

PR #1558

**Comment by boaks** [comment]
FMPOV, that's done with the work in master/3.0.

Other opinions?
